### PR TITLE
Document dialogue data structures

### PR DESCRIPTION
**Pull request by justinvdm**
The plan is to define dialogue, sequence and block [json schema](http://json-schema.org/)s using yaml files, and include these in the docs as a way of documenting how these data structures should look.

This approach seems better than maintaining separate documentation for these structures along with the schema:
- the structure of a dialogue is complex enough to be difficult to document using sphinx alone
- separate docs + schema approach allows room for outdated docs

### CODE REVIEW

**Comment by justinvdm** [comment]
Ready for review.

**Comment by justinvdm** [comment]
The documented structures differ slightly from how the structures look in numl. We will need to change numl to match this:
- `id` identifies blocks and sequences rather than `name`. I think `id` makes more sense as a field name there anyways.
- we need to flatten the `properties` data structure for dialogues, sequences and blocks in numl. For the case of dialogues, for example, from the API's point of view, separating `title` and `sequences` from the rest of the fields (for example, the `id` field) makes no sense. Having all properties be at the same level also seems to make more sense for numl.

We need to think a bit about the fact that ids  for blocks and sequences are provided by the client as identifiers rather than falling back to generated identifiers. For numl, this means we are relying on the ids people provide as identifiers for storing and looking up blocks and sequences. Blocks and sequences only need to be unique to their siblings, and this is easy for us to validate, so I think this is safe enough. This approach seems more ideal than the alternatives I could think up, happy to hear other suggestions for this.

File: schemas/block.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    description: |
+      Identifies the type of block. Valid values depend on which block types
+      have been registered with a particular instance of the API.
+
+  title:
+    type: text
+    description: |
+      A human-readable title describing the block.
+
+additionalProperties: false
```
Comment (hodgestar) [inline schemas/block.yml:28]: Couldn't block have additional properties specific to their type?

**Comment by hodgestar** [comment]
Thoughts:
- Happy with the `name -> id` change.
- Happy for `id`, `title` and `children` (so `sequences`, `blocks`) to be on the same level.
- A bit less sure about adding arbitrary other properties on the same level. Aren't some of them already on the same level in numl, at least for blocks?

**Comment by justinvdm** [comment]
> A bit less sure about adding arbitrary other properties on the same level. Aren't some of them already on the same level in numl, at least for blocks?

Yep, at the moment, in numl, blocks look like `{title: ..., properties: ...}` and sequences look like `{title: ..., blocks: ..., properties: ...}`. Separating `title`, `properties` and the children property from the rest of the properties seemed more like a side-effect of the way those structures **look** in numl. Your [comment](https://github.com/praekelt/numi-api/pull/4#discussion_r57976456) about arbitrary properties for different block types is a good point, made me realise `properties` should maybe be reserved for properties only relevant to a particular block type.

 How about a structure like this for blocks:

``` js
{
  id: ...,
  title: ...,
  type: ...
  properties: {/* properties relevant to a particular block type */}
}
```

and this for sequences (we no longer need a `properties` object here):

``` js
{
  id: ...,
  title: ...,
  blocks: [/* block definitions */]
}
```

For numl, it does mean special casing the `id` and `type` properties and moving them to the top level, but I don't think there's any problems with that approach.

**Comment by justinvdm** [comment]
@hodgestar any thoughts on this part of my first comment (I fixed it up a bit to read better)? 

> We need to think a bit about the fact that ids  for blocks and sequences are provided by the client as identifiers rather than falling back to generated identifiers. For numl, this means we are relying on the ids people provide as identifiers for storing and looking up blocks and sequences. Blocks and sequences only need to be unique to their siblings, and this is easy for us to validate, so I think this is safe enough. This approach seems more ideal than the alternatives I could think up, happy to hear other suggestions for this.

**Comment by justinvdm** [comment]
@hodgestar provided you are happy with my suggestion, ready for re-review.

File: schemas/block.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+required:
+  - id
+  - type
+
+properties:
+  id:
+    $ref: '#/definitions/symbol'
+    description: |
+      The block's identifier. Must be generated by the client.
```
Comment (hodgestar) [inline schemas/block.yml:9]: I'm fine with this being client generated.

**Comment by hodgestar** [comment]
:+1:

**Pull request merged**